eclair-core/src/main/scala/fr/acinq/eclair/io/ReconnectionTask.scala

```diff
@@ -0,0 +1,189 @@
+        // The random randomizeDelay does two things:
+        // - it adds a minimum delay. This delay is important in the case of a first connection to a new peer which advertises
+        // a public address. Right after the peer and this reconnection actor will be created, there will be a race between
+        // this automated connection task that uses data from network db, and the Peer.Connect command that may use the same
+        // address or a different one. This delay will cause the automated connection task to lose the race and prevent
+        // unnecessary parallel connections
+        // - it also adds some randomization. This is important in the case of a reconnection. If both peers have a public
+        // address, they may attempt to simultaneously connect back to each other, which could result in reconnection loop,
+        // given that each new connection will cause the previous one to be killed.
+        val initialDelay = randomizeDelay(nodeParams.initialRandomReconnectDelay)
+        val firstNextReconnectionDelay = previousState match {
+          case Peer.INSTANTIATING =>
+            // When restarting, we will ~immediately reconnect, but then:
+            // - we don't want all the subsequent reconnection attempts to be synchronized (herd effect)
+            // - we don't want to go through the exponential backoff delay, because we were offline, not them, so there is no
+            // reason to eagerly retry
+            // That's why we set the next reconnection delay to a random value between MAX_RECONNECT_INTERVAL/2 and MAX_RECONNECT_INTERVAL.
+            nodeParams.maxReconnectInterval.minus(Random.nextInt(nodeParams.maxReconnectInterval.toSeconds.toInt / 2).seconds)
+          case Peer.CONNECTED =>
+            log.info("peer is disconnected")
+            nextReconnectionDelay(initialDelay, nodeParams.maxReconnectInterval)
+        }
+        setTimer(RECONNECT_TIMER, TickReconnect, initialDelay, repeat = false)
+        goto(WAITING) using WaitingData(firstNextReconnectionDelay)
+      } else {
+        stay
+      }
+
+    case Event(FSM.Transition(_, Peer.DISCONNECTED, Peer.CONNECTED), _) =>
+      log.info("peer is connected")
+      stay
+  }
+
+  whenUnhandled {
+    case Event(_: Terminated, _) => stay
+
+    case Event(TickReconnect, _) => stay
+
+    case Event(FSM.Transition(_, Peer.INSTANTIATING, Peer.INSTANTIATING), _) => stay // instantiation transition
+
+    case Event(Peer.Connect(_, hostAndPort_opt), _) =>
+      // this happens completely independently of the automated reconnection process
+      // if we are already connecting/connected, the peer will kill any duplicate connections
+      hostAndPort_opt
+        .map(hostAndPort2InetSocketAddress)
+        .orElse(getPeerAddressFromDb(nodeParams.db.peers, nodeParams.db.network, remoteNodeId)) match {
+        case Some(address) => connect(address, Some(sender))
+        case None => sender ! "no address found"
+      }
+      stay
```

Comment:
Shouldn't we go to CONNECTING? If we do that you probably wouldn't need the `case Event(_: Terminated, _)` in the `whenUnhandled` (or am I missing something?).

Reply:
No, that's the tricky part (see my comment https://github.com/ACINQ/eclair/pull/1347#issuecomment-600803658). The user-requested connect is treated independently of the regular automated connection loop. Feels a bit hackish at first but it makes sense.

Reply:
Ok, I get it, this is indeed a bit subtle, maybe worth a comment above that `stay`?

Reply:
The comment is there already (lines 124,125), not sure what I can add to make it clearer

Reply:
I would add that this explains why we don't modify our state (it's far from obvious IMO).
Something like:
```
// manual connection requests happen completely independently of the automated reconnection process;
// we initiate a connection but don't modify our state.
// if we are already connecting/connected, the peer will kill any duplicate connections
```

Reply:
Done with 5106463